In [0]:
#install category_encoders first
!pip install category_encoders

     |████████████████████████████████| 102kB 2.2MB/s 


We are going to apply the different encoding techniques on big mart sales data.
 

---
Things to learn - 

*   Indentifying data type as ordinal,nominal and continuous.
*   Applying different types of encoding. 
*   Challenges with different encoding techniques.
*   Choosing the appropriate encoding techniques.




In [0]:
import pandas as pd #import pandas
import numpy as np #import numpy
from sklearn.preprocessing import LabelEncoder  #importing LabelEncoder
import warnings
warnings.filterwarnings("ignore")

In [0]:
train = pd.read_csv('/content/drive/My Drive/Feature Encoding/feature_en/Feature_encoding/train_bm.csv')

In [0]:
#check the head of dataset
train.head(5)

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

In [0]:
#check the size of the dataset
print('Data has {} Number of rows'.format(train.shape[0]))
print('Data has {} Number of columns'.format(train.shape[1]))

Data has 8523 Number of rows
Data has 12 Number of columns


In [0]:
#check the information of the dataset
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


As we can see here, we have 7 categorical variables and 5 numeric variables.
The first task is to identify these categorical variables as nominal or ordinal.

In [0]:
#let's keep our categorical variables in one table
cat_data = train[['Item_Identifier','Item_Fat_Content','Item_Type','Outlet_Identifier','Outlet_Size','Outlet_Location_Type','Outlet_Type']]

In [0]:
cat_data.head()   #check the head of categorical data

Item_Identifier Item_Fat_Content  ... Outlet_Location_Type        Outlet_Type
0           FDA15          Low Fat  ...               Tier 1  Supermarket Type1
1           DRC01          Regular  ...               Tier 3  Supermarket Type2
2           FDN15          Low Fat  ...               Tier 1  Supermarket Type1
3           FDX07          Regular  ...               Tier 3      Grocery Store
4           NCD19          Low Fat  ...               Tier 3  Supermarket Type1

[5 rows x 7 columns]

In [0]:
cat_data.apply(lambda x: x.nunique()) #check the number of unique values in each column

Item_Identifier         1559
Item_Fat_Content           5
Item_Type                 16
Outlet_Identifier         10
Outlet_Size                3
Outlet_Location_Type       3
Outlet_Type                4
dtype: int64

Now think which encoding technique can we apply here.


*   First thought would be to apply one hot encoding on features which has 3-5 unique categories. 
*   But what if there is some kind of ordering present between them. So firstly we should identify the nominal and ordinal variable
*   Let's check one by one



In [0]:
#check the top 10 frequency in Item_Identifier
cat_data['Item_Identifier'].value_counts().head(10)

FDG33    10
FDW13    10
FDP25     9
FDQ40     9
FDX20     9
FDW49     9
FDX31     9
NCF42     9
FDT07     9
DRN47     9
Name: Item_Identifier, dtype: int64

The values in Item_Identifier has no ordering as we can see. These are nominal categorical variable. 

The first column has 1559 unique values. If we try to do one hot encoding here we will have 1558 new features. We cannot feed in these many features in our model. It will make our model complex and it will reduce the model accuracy.

In [0]:
pd.get_dummies(cat_data['Item_Identifier'],drop_first=True)  #applying one hot encoding

DRA24  DRA59  DRB01  DRB13  DRB24  ...  NCZ30  NCZ41  NCZ42  NCZ53  NCZ54
0         0      0      0      0      0  ...      0      0      0      0      0
1         0      0      0      0      0  ...      0      0      0      0      0
2         0      0      0      0      0  ...      0      0      0      0      0
3         0      0      0      0      0  ...      0      0      0      0      0
4         0      0      0      0      0  ...      0      0      0      0      0
...     ...    ...    ...    ...    ...  ...    ...    ...    ...    ...    ...
8518      0      0      0      0      0  ...      0      0      0      0      0
8519      0      0      0      0      0  ...      0      0      0      0      0
8520      0      0      0      0      0  ...      0      0      0      0      0
8521      0      0      0      0      0  ...      0      0      0      0      0
8522      0      0      0      0      0  ...      0      0      0      0      0

[8523 rows x 1558 columns]

As expected from a single feature now we have 1558 features. So it's a bad idea to apply one hot encoding here. We should not apply one hot encoding when there are too many categories.

So one hot encoding has failed us here. Now for rescue we move to LabelEncoding but we are very much aware that if we apply label encoding on a feature it assigns a natural ranking to the categories alphabatically. So we cannot apply Label encoding as well. 

So we have 1 thing left (Binary Encoding) that we have learnt previously. Let's apply it and see what we get.

In [0]:
#apply binary encoding on Item_Identifier
import category_encoders as ce                              #import category_encoders
encoder = ce.BinaryEncoder(cols=['Item_Identifier'])        #create instance of binary enocder
df_binary = encoder.fit_transform(cat_data)                 #fit and tranform on cat_data
df_binary.head(5)

Item_Identifier_0  Item_Identifier_1  ...  Outlet_Location_Type        Outlet_Type
0                  0                  0  ...                Tier 1  Supermarket Type1
1                  0                  0  ...                Tier 3  Supermarket Type2
2                  0                  0  ...                Tier 1  Supermarket Type1
3                  0                  0  ...                Tier 3      Grocery Store
4                  0                  0  ...                Tier 3  Supermarket Type1

[5 rows x 18 columns]

In [0]:
#check the columns
df_binary.columns

Index(['Item_Identifier_0', 'Item_Identifier_1', 'Item_Identifier_2',
       'Item_Identifier_3', 'Item_Identifier_4', 'Item_Identifier_5',
       'Item_Identifier_6', 'Item_Identifier_7', 'Item_Identifier_8',
       'Item_Identifier_9', 'Item_Identifier_10', 'Item_Identifier_11',
       'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

Binary encoder has given us 11 new feature which is way less than we were getting from one hot encoding. So we have been rescued here by Binary Encoding.

We have applied binary encoding but it doesn't provide us any intution as how these new features are made. All we know is by using binary encoding Here the labels are firstly encoded ordinal and then they are converted into binary codes. Then the digits from that binary string are converted into different features.

There are other intutive measures to reduce the features. We will look at them later.

##Encoding Item_Fat_Content

In [0]:
#check the unique values 
cat_data['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

Here we have 5 unique values but if we look at them closely there are only 2 unique values. Low Fat and Regular, others are just short forms for them or are in small letters.

In [0]:
low_fat = ['LF','low fat']
cat_data['Item_Fat_Content'].replace(low_fat,'Low Fat',inplace = True) #replace 'LF' and 'low fat' with 'Low Fat'
cat_data['Item_Fat_Content'].replace('reg','Regular',inplace = True)   #Replace 'reg' with regular

In [0]:
cat_data['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular'], dtype=object)

Here we have 2 categories in Item_Fat_Content and we have some ordering between the. Low Fat will have less Fat content than the regular Fat. So it is a ordinal variable. 

In [0]:
#Apply LabelEncoder
le = LabelEncoder()
cat_data['Item_Fat_Content_temp'] = le.fit_transform(cat_data['Item_Fat_Content'])
print(cat_data['Item_Fat_Content'].head())
print(cat_data['Item_Fat_Content_temp'].head())

0    Low Fat
1    Regular
2    Low Fat
3    Regular
4    Low Fat
Name: Item_Fat_Content, dtype: object
0    0
1    1
2    0
3    1
4    0
Name: Item_Fat_Content_temp, dtype: int64


Here we only had 2 categories 'Low Fat' and 'Regular' so using LabelEncoding has worked here. It has mapped :-


*   Low Fat ------- 0
*   Regular ------- 1



Here the natural ranking of alphabets has worked but every time you are not this lucky.

##We can use map to do ordinal encoding

In [0]:
#prepare a dict to map
mapping = {'Low Fat' : 0,'Regular': 1} #map Low Fat as 0 and Regular as 1
cat_data['Item_Fat_Content_temp1'] = cat_data['Item_Fat_Content'].map(mapping)
cat_data['Item_Fat_Content_temp1'].head()

0    0
1    1
2    0
3    1
4    0
Name: Item_Fat_Content_temp1, dtype: int64

It is useful when we have ordering in our categories.

# Use Pandas pd.factorize method.
It does the nominal encoding based on the order in which the categories apper. If Low Fat is at index 0 then it will be encoded as 0 Regular as 1 and vice versa.

In [0]:
factorized,index = pd.factorize(cat_data['Item_Fat_Content'])  #using pd.factorize it gives us factorized array and index values
print(factorized)
print(index)

[0 1 0 ... 0 1 0]
Index(['Low Fat', 'Regular'], dtype='object')


In this Notebook we have seen 2 new encoding techniques.


*   Mapping 
*   pd.factorize


---
We have seen the usage of different methods, their advantages and disadvantages.



In [0]:
#Let's look at item type column
print(cat_data['Item_Type'].nunique())  #check number of unique values
print(cat_data['Item_Type'].unique())   #check the unique values

16
['Dairy' 'Soft Drinks' 'Meat' 'Fruits and Vegetables' 'Household'
 'Baking Goods' 'Snack Foods' 'Frozen Foods' 'Breakfast'
 'Health and Hygiene' 'Hard Drinks' 'Canned' 'Breads' 'Starchy Foods'
 'Others' 'Seafood']


And we don't Have any ordering between them. So we have to apply ordinal encoding technique. i Leave it upto you to decide which technique to apply and we will have look at other techniques in our next Notebook.